File: frontend/src/api/health.ts
import { apiRequest, type ApiValue } from "./client";

export type HealthRecordPayload = Record<string, ApiValue> & {
  height_cm?: number;
  weight_kg?: number;
  bmi?: number;
  waist_cm?: number;
  systolic_bp?: number;
  diastolic_bp?: number;
  fasting_glucose?: number;
  hba1c?: number;
  total_cholesterol?: number;
  ldl_cholesterol?: number;
  hdl_cholesterol?: number;
  triglyceride?: number;
  ast?: number;
  alt?: number;
  gamma_gtp?: number;
  creatinine?: number;
  egfr?: number;
  hemoglobin?: number;
  has_diabetes?: boolean;
  has_obesity?: boolean;
  has_dyslipidemia?: boolean;
  has_hypertension?: boolean;
  occupation_code?: string;
  family_htn?: string;
  family_dm?: string;
  family_dyslipidemia?: string;
  smoking_status?: string;
  drinking_frequency?: string;
  drinking_amount?: string;
  walking_days_per_week?: number;
  strength_days_per_week?: number;
  sleep_hours?: number;
  source?: "MANUAL" | "OCR" | "PROFILE" | "ANALYSIS_PREP";
  measured_at: string;
};

export async function listHealthRecords<T>(): Promise<T> {
  return apiRequest<T>("/health/records");
}

export async function getLatestHealthRecord<T>(): Promise<T> {
  return apiRequest<T>("/health/records/latest");
}

export async function createHealthRecord<T>(payload: HealthRecordPayload): Promise<T> {
  return apiRequest<T>("/health/records", { method: "POST", body: payload });
}

export async function updateHealthRecord<T>(
  recordId: number,
  payload: Partial<HealthRecordPayload>,
): Promise<T> {
  return apiRequest<T>(`/health/records/${recordId}`, { method: "PATCH", body: payload });
}

export async function deleteHealthRecord<T>(recordId: number): Promise<T> {
  return apiRequest<T>(`/health/records/${recordId}`, { method: "DELETE" });
}

export async function getAnalysisReadiness<T>(): Promise<T> {
  return apiRequest<T>("/health/analysis-readiness");
}

File: app/dtos/health.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field

from app.dtos.base import BaseSerializerModel


class HealthRecordCreateRequest(BaseModel):
    height_cm: Decimal | None = None
    weight_kg: Decimal | None = None
    bmi: Decimal | None = None
    waist_cm: Decimal | None = None
    systolic_bp: int | None = None
    diastolic_bp: int | None = None
    fasting_glucose: int | None = None
    hba1c: Decimal | None = None
    total_cholesterol: int | None = None
    ldl_cholesterol: int | None = None
    hdl_cholesterol: int | None = None
    triglyceride: int | None = None
    has_diabetes: bool | None = None
    has_obesity: bool | None = None
    has_dyslipidemia: bool | None = None
    has_hypertension: bool | None = None
    occupation_code: str | None = None
    family_htn: str | None = None
    family_dm: str | None = None
    family_dyslipidemia: str | None = None
    smoking_status: str | None = None
    drinking_frequency: str | None = None
    drinking_amount: str | None = None
    walking_days_per_week: int | None = Field(default=None, ge=0, le=7)
    strength_days_per_week: int | None = Field(default=None, ge=0, le=7)
    sleep_hours: Decimal | None = None
    measured_at: datetime


class HealthRecordUpdateRequest(BaseModel):
    height_cm: Decimal | None = None
    weight_kg: Decimal | None = None
    bmi: Decimal | None = None
    waist_cm: Decimal | None = None
    systolic_bp: int | None = None
    diastolic_bp: int | None = None
    fasting_glucose: int | None = None
    hba1c: Decimal | None = None
    total_cholesterol: int | None = None
    ldl_cholesterol: int | None = None
    hdl_cholesterol: int | None = None
    triglyceride: int | None = None
    has_diabetes: bool | None = None
    has_obesity: bool | None = None
    has_dyslipidemia: bool | None = None
    has_hypertension: bool | None = None
    occupation_code: str | None = None
    family_htn: str | None = None
    family_dm: str | None = None
    family_dyslipidemia: str | None = None
    smoking_status: str | None = None
    drinking_frequency: str | None = None
    drinking_amount: str | None = None
    walking_days_per_week: int | None = Field(default=None, ge=0, le=7)
    strength_days_per_week: int | None = Field(default=None, ge=0, le=7)
    sleep_hours: Decimal | None = None
    measured_at: datetime | None = None


class HealthRecordResponse(BaseSerializerModel):
    id: int
    user_id: int
    height_cm: Decimal | None
    weight_kg: Decimal | None
    bmi: Decimal | None
    waist_cm: Decimal | None
    systolic_bp: int | None
    diastolic_bp: int | None
    fasting_glucose: int | None
    hba1c: Decimal | None
    total_cholesterol: int | None
    ldl_cholesterol: int | None
    hdl_cholesterol: int | None
    triglyceride: int | None
    has_diabetes: bool | None
    has_obesity: bool | None
    has_dyslipidemia: bool | None
    has_hypertension: bool | None
    occupation_code: str | None
    family_htn: str | None
    family_dm: str | None
    family_dyslipidemia: str | None
    smoking_status: str | None
    drinking_frequency: str | None
    drinking_amount: str | None
    walking_days_per_week: int | None
    strength_days_per_week: int | None
    is_smoker: bool | None
    drinks_alcohol: bool | None
    exercise_days_per_week: int | None
    sleep_hours: Decimal | None
    measured_at: datetime
    created_at: datetime
    updated_at: datetime


class HealthRecordListResponse(BaseModel):
    items: list[HealthRecordResponse]
    total: int


class HealthAnalysisReadinessResponse(BaseModel):
    is_ready: bool
    basic_ready: bool | None = None
    precision_ready: bool | None = None
    latest_health_record_id: int | None = None
    missing_fields: list[str]
    missing_basic_fields: list[str] = []
    missing_precision_fields: list[str] = []
    message: str
